### ISSUE TITLE
404 error trying to install a Bioconductor package inside the official Bioconductor docker container

### ISSUE THREAD

**Issue by tomsing1** [rstudio/renv#1013]
### tl;dr:

- `renv::install("bioc::S4Vectors")` fails with a `404` error inside the Bioconductor docker container.
- The URL `https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0.tar.gz` doesn't seem to exist.
- Perhaps `renv` should be looking for `https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0_R_x86_64-pc-linux-gnu.tar.gz` instead?

Dear renv team,

I am running R & RStudio inside the latest Bioconductor docker image (RELEASE_3_15):

```bash
docker run \
  --rm \
  -ti \
  -p 127.0.0.1:8787:8787 \
  -e DISABLE_AUTH=true bioconductor/bioconductor_docker:RELEASE_3_15
````

Then, inside the R session, I run the following code to install `renv` inside the container:

```r
install.packages("renv")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.rstudio.com/cran/__linux__/focal/latest/src/contrib/renv_0.15.4.tar.gz'
Content type 'binary/octet-stream' length 1846373 bytes (1.8 MB)
==================================================
downloaded 1.8 MB

* installing *binary* package ‘renv’ ...
* DONE (renv)

The downloaded source packages are in
	‘/tmp/RtmpmpQFiE/downloaded_packages’
```

Next, I create a new directory and initiate a new `renv` project:

```r
dir.create("~/test")
setwd("~/test/")
renv::init(bioconductor = "3.15")  # restarts the R session

* Using Bioconductor version '3.15'.
* Initializing project ...
* Discovering package dependencies ... Done!
* Copying packages into the cache ... Done!
The following package(s) will be updated in the lockfile:

# RSPM ===============================
- renv   [* -> 0.15.4]

* Lockfile written to '~/test/renv.lock'.
> 

Restarting R session...
* Project '~/test' loaded. [renv 0.15.4]
```

Once the R session has restarted, I try to install the S4Vectors Bioconductor package.
But unfortunately, the installation fails with a `404` error:

```r
library(renv)
install("bioc::S4Vectors")

Retrieving 'https://packagemanager.rstudio.com/cran/__linux__/focal/latest/src/contrib/BiocManager_1.30.18.tar.gz' ...
	OK [downloaded 353 Kb in 0.7 secs]
Installing BiocManager [1.30.18] ...
	OK [installed binary]
Moving BiocManager [1.30.18] into the cache ...
	OK [moved to cache in 0.39 milliseconds]
Retrieving 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0.tar.gz' ...
Error: failed to retrieve 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0.tar.gz' [error code 22]
In addition: Warning message:
curl: (22) The requested URL returned error: 404  
```

It seems that the url `https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0.tar.gz` renv tries to access doesn't exist.

Any ideas how I can fix this? 😕  
Many thanks for any pointers! 

### Attempted workaround

To work around this problem, I used the `BiocManager::install` function instead. Interestingly, 
that retrieves the `S4Vectors` from a slightly different URL: `https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0_R_x86_64-pc-linux-gnu.tar.gz`.

```r
BiocManager::install("S4Vectors")
'getOption("repos")' replaces Bioconductor standard repositories, see '?repositories'
for details

replacement repositories:
    CRAN: https://packagemanager.rstudio.com/cran/__linux__/focal/latest

Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.0 (2022-04-22)
Installing package(s) 'BiocVersion', 'S4Vectors'
also installing the dependency ‘BiocGenerics’

trying URL 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/BiocGenerics_0.42.0_R_x86_64-pc-linux-gnu.tar.gz'
Content type 'application/x-tar' length 593206 bytes (579 KB)
==================================================
downloaded 579 KB

trying URL 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/BiocVersion_3.15.2_R_x86_64-pc-linux-gnu.tar.gz'
Content type 'application/x-tar' length 5402 bytes
==================================================
downloaded 5402 bytes

trying URL 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0_R_x86_64-pc-linux-gnu.tar.gz'
Content type 'application/x-tar' length 2211356 bytes (2.1 MB)
==================================================
downloaded 2.1 MB

* installing *binary* package ‘BiocGenerics’ ...
* DONE (BiocGenerics)
* installing *binary* package ‘BiocVersion’ ...
* DONE (BiocVersion)
* installing *binary* package ‘S4Vectors’ ...
* DONE (S4Vectors)

The downloaded source packages are in
	‘/tmp/RtmpLFl4ZT/downloaded_packages’
```

Next, I snapshot the package:

```r
> snapshot(packages = "S4Vectors")
The following package(s) will be updated in the lockfile:

# Bioconductor =======================
- BiocGenerics   [* -> 0.42.0]
- BiocVersion    [* -> 3.15.2]
- S4Vectors      [* -> 0.34.0]

# RSPM ===============================
- renv           [0.15.4 -> *]
- BiocManager    [* -> 1.30.18]

Do you want to proceed? [y/N]: y
* Lockfile written to '~/test/renv.lock'.
```

The `renv.lock` file now contains:

```json
{
  "R": {
    "Version": "4.2.0",
    "Repositories": [
      {
        "Name": "CRAN",
        "URL": "https://packagemanager.rstudio.com/cran/latest"
      }
    ]
  },
  "Bioconductor": {
    "Version": "3.15"
  },
  "Packages": {
    "BiocGenerics": {
      "Package": "BiocGenerics",
      "Version": "0.42.0",
      "Source": "Bioconductor",
      "git_url": "https://git.bioconductor.org/packages/BiocGenerics",
      "git_branch": "RELEASE_3_15",
      "git_last_commit": "3582d47",
      "git_last_commit_date": "2022-04-26",
      "Hash": "37348ee784e82c0a6b650146275b459d",
      "Requirements": []
    },
    "BiocManager": {
      "Package": "BiocManager",
      "Version": "1.30.18",
      "Source": "Repository",
      "Repository": "RSPM",
      "Hash": "b1a93bed5debda5775636086fdca017b",
      "Requirements": []
    },
    "BiocVersion": {
      "Package": "BiocVersion",
      "Version": "3.15.2",
      "Source": "Bioconductor",
      "git_url": "https://git.bioconductor.org/packages/BiocVersion",
      "git_branch": "master",
      "git_last_commit": "818ab03",
      "git_last_commit_date": "2022-03-29",
      "Hash": "48e545a2fa24a967de72b1114bfe96ac",
      "Requirements": []
    },
    "S4Vectors": {
      "Package": "S4Vectors",
      "Version": "0.34.0",
      "Source": "Bioconductor",
      "git_url": "https://git.bioconductor.org/packages/S4Vectors",
      "git_branch": "RELEASE_3_15",
      "git_last_commit": "f590de3",
      "git_last_commit_date": "2022-04-26",
      "Hash": "90677cc888563927cecacd4067ccad45",
      "Requirements": [
        "BiocGenerics"
      ]
    }
  }
}
```

Finally, I try to use this `renv.lock` file to restore the environment in a new directory.
But unfortunately, `renv` tries to access a non-existing URL again - this time for the `BiocGenerics` Bioconductor package:

```r
dir.create("~/test2")
setwd("~/test2")
init() # restarts the R session
```

```r
restore(lockfile = "~/test/renv.lock")

The following package(s) will be updated:

# Bioconductor =======================
- BiocGenerics   [* -> 0.42.0]
- BiocVersion    [* -> 3.15.2]
- S4Vectors      [* -> 0.34.0]

# RSPM ===============================
- BiocManager    [* -> 1.30.18]

Do you want to proceed? [y/N]: y
Installing BiocManager [1.30.18] ...
	OK [linked cache]
* Querying repositories for available source packages ... Done!
Retrieving 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/BiocGenerics_0.42.0.tar.gz' ...
Warning: failed to retrieve 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/BiocGenerics_0.42.0.tar.gz' [error code 22]
Error: failed to retrieve package 'BiocGenerics@0.42.0'
In addition: Warning messages:
1: curl: (22) The requested URL returned error: 404  
2: failed to retrieve 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/BiocGenerics_0.42.0.tar.gz' [error code 22] 
```

## Session Information

```r
.libPaths()
[1] "/home/rstudio/test/renv/library/R-4.2/x86_64-pc-linux-gnu"
[2] "/usr/local/lib/R/library"  

sessionInfo()
R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 20.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
[1] renv_0.15.4

loaded via a namespace (and not attached):
[1] BiocManager_1.30.18 compiler_4.2.0      tools_4.2.0  
```

**Comment by tomsing1** [comment]
A few more observations and a *working workaround*:

1.  Setting `renv.config.rspm.enabled` to FALSE does **not** resolve the issue:

```r
> options(renv.config.rspm.enabled = FALSE)
> install("bioc::S4Vectors")
Retrieving 'https://packagemanager.rstudio.com/cran/__linux__/focal/latest/src/contrib/BiocManager_1.30.18.tar.gz' ...
	OK [downloaded 353 Kb in 1 secs]
Installing BiocManager [1.30.18] ...
	OK [installed binary]
Moving BiocManager [1.30.18] into the cache ...
	OK [moved to cache in 0.41 milliseconds]
Retrieving 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0.tar.gz' ...
Error: failed to retrieve 'https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker/src/contrib/S4Vectors_0.34.0.tar.gz' [error code 22]
In addition: Warning message:
curl: (22) The requested URL returned error: 404  
```

2.  Different Bioc repositories in the bioconductor and rocker/rstudio docker images

The installation **succeeds** in the latest `rocker/rstudio` container. I noticed that the `BiocManager::repositories()` command returns different output in the two containers:

In the `bioconductor` docker container:

```r
BiocManager::repositories()
'getOption("repos")' replaces Bioconductor standard repositories, see '?repositories'
for details

replacement repositories:
    CRAN: https://packagemanager.rstudio.com/cran/__linux__/focal/latest

                                                                 BioCcontainers 
"https://bioconductor.org/packages/3.15/container-binaries/bioconductor_docker" 
                                                                       BioCsoft 
                                  "https://bioconductor.org/packages/3.15/bioc" 
                                                                        BioCann 
                       "https://bioconductor.org/packages/3.15/data/annotation" 
                                                                        BioCexp 
                       "https://bioconductor.org/packages/3.15/data/experiment" 
                                                                  BioCworkflows 
                             "https://bioconductor.org/packages/3.15/workflows" 
                                                                      BioCbooks 
                                 "https://bioconductor.org/packages/3.15/books" 
                                                                           CRAN 
               "https://packagemanager.rstudio.com/cran/__linux__/focal/latest" 
```
 
but in `rocker/rstudio` container:

```r
BiocManager::repositories()

                                                        BioCsoft 
                   "https://bioconductor.org/packages/3.15/bioc" 
                                                         BioCann 
        "https://bioconductor.org/packages/3.15/data/annotation" 
                                                         BioCexp 
        "https://bioconductor.org/packages/3.15/data/experiment" 
                                                   BioCworkflows 
              "https://bioconductor.org/packages/3.15/workflows" 
                                                       BioCbooks 
                  "https://bioconductor.org/packages/3.15/books" 
                                                            CRAN 
"https://packagemanager.rstudio.com/cran/__linux__/focal/latest" 
```

When I pass the latter repositories to the `renv::install` command the `bioconductor` container manually, the installation succeeds (from source):

```r
bioc_repos <- c(BioCsoft =
                "https://bioconductor.org/packages/3.15/bioc", 
                BioCann =
                "https://bioconductor.org/packages/3.15/data/annotation",
                BioCexp =
                "https://bioconductor.org/packages/3.15/data/experiment",
                BioCworkflows =
                "https://bioconductor.org/packages/3.15/workflows",
                BioCbooks =
                "https://bioconductor.org/packages/3.15/books",
                CRAN =
                "https://packagemanager.rstudio.com/cran/__linux__/focal/latest" )

install("bioc::S4Vectors", repos = bioc_repos)

Retrieving 'https://bioconductor.org/packages/3.15/bioc/src/contrib/S4Vectors_0.34.0.tar.gz' ...
	OK [downloaded 604.1 Kb in 0.7 secs]
Retrieving 'https://bioconductor.org/packages/3.15/bioc/src/contrib/BiocGenerics_0.42.0.tar.gz' ...
	OK [downloaded 44.9 Kb in 0.2 secs]
Installing BiocGenerics [0.42.0] ...
	OK [built from source]
Moving BiocGenerics [0.42.0] into the cache ...
	OK [moved to cache in 0.41 milliseconds]
Installing S4Vectors [0.34.0] ...
	OK [built from source]
Moving S4Vectors [0.34.0] into the cache ...
	OK [moved to cache in 0.38 milliseconds]
```

So it seems that the default repositories renv uses inside the `bioconductor` container may be the problem.
Maybe I need to update the repositories permanently? What's the best way to do that?

Many thanks for any pointers!

**Comment by kevinushey** [comment]
I think this should be fixed via https://github.com/rstudio/renv/commit/7cee0d2d6966dc3d18ff30af327119ef490ae40b. Would you be able to upgrade to the development version of `renv` and let me know?

**Comment by tomsing1** [comment]
Yay, with `renv` version `0.15.4-45` the `S4Vectors` installation succeeds in the bioconductor image as well.
Many thanks for looking into this so quickly!

**Comment by kevinushey** [comment]
Awesome, thank you for testing! I'll think about putting a CRAN release together soon.
